larsim/GenericCRT/GenericCRT.cxx

```diff
@@ -0,0 +1,150 @@
+         //if channel ID is not in the set yet, add it
+      AuxDetChanNumbers.push_back(HitIter.GetID());
+        }//
+
+  }
+
+return AuxDetChanNumbers.size();
+
+}
+
+
+
+sim::AuxDetSimChannel const sim::GenericCRTUtility::GetAuxDetSimChannelByNumber( const std::vector<sim::AuxDetHit> &InputHitVector,unsigned int inputchannel)
+{
+        std::vector<sim::AuxDetIDE> IDEvector;
+       IDEvector.resize(0);
+        //loop over sim::AuxDetHits and assign them to AuxDetSimChannels.
+
+        size_t ad_id_no = 9999;
+        size_t ad_sen_id_no = 9999;
+
+        for(auto AuxDetHitIter : InputHitVector)
+        {
+
+            unsigned int channel = AuxDetHitIter.GetID();
+            double xcoordinate = (AuxDetHitIter.GetEntryX() + AuxDetHitIter.GetExitX())/2.0;
+            double ycoordinate = (AuxDetHitIter.GetEntryY() + AuxDetHitIter.GetExitY())/2.0;
+            double zcoordinate = (AuxDetHitIter.GetEntryZ() + AuxDetHitIter.GetExitZ())/2.0;
+            double worldPos[3] = {xcoordinate,ycoordinate,zcoordinate};
+            fGeo->FindAuxDetSensitiveAtPosition(worldPos, ad_id_no, ad_sen_id_no, 0.0001);
+            if(channel == inputchannel)   // this is the channel we want.
+            {
+                sim::AuxDetIDE tempIDE;
+                CopyAuxDetHittoAuxDetIDE(AuxDetHitIter,tempIDE);
+
+                std::vector<sim::AuxDetIDE>::iterator IDEitr
+                = std::find(IDEvector.begin(), IDEvector.end(), tempIDE);
+
+                if(IDEitr != IDEvector.end()){ //If trackID is already in the map, update it
+                    //Andrzej's note - following logic from AuxDetReadout in Legacy, but why are the other paremeters getting overwritten like that?
+                    IDEitr->energyDeposited += tempIDE.energyDeposited;
+                    IDEitr->exitX            = tempIDE.exitX;
+                    IDEitr->exitY            = tempIDE.exitY;
+                    IDEitr->exitZ            = tempIDE.exitZ;
+                    IDEitr->exitT            = tempIDE.exitT;
+                    IDEitr->exitMomentumX    = tempIDE.exitMomentumX;
+                    IDEitr->exitMomentumY    = tempIDE.exitMomentumY;
+                    IDEitr->exitMomentumZ    = tempIDE.exitMomentumZ;
+                    }
+                    else{  //if trackID is not in the set yet, add it
+                        IDEvector.push_back(std::move(tempIDE));
+                    }//else
+
+
+            } // end if the AuxDetHit channel checks out.
+
+        } // end main loop on AuxDetHit
+
+        //push back the AuxDetSimChannel Vector.
+        //TODO check the last parameter values.
+       const sim::AuxDetSimChannel adsc=sim::AuxDetSimChannel(ad_id_no, std::move(IDEvector), ad_sen_id_no);
+
+  return adsc;
+}
+
+
+
+
+void sim::GenericCRTUtility::FillAuxDetSimChannels(const std::vector<sim::AuxDetHit> &InputHitVector, std::vector<sim::AuxDetSimChannel> *AuxDetVector)
+{
+
```

Comment:
A lot of work going on in this function.  It can be simplified:

```c++
std::vector<sim::AuxDetSimChannel>
sim::GenericCRTUtility::GetAuxDetSimChannels(const std::vector<sim::AuxDetHit>& InputHitVector) const
{
  auto const auxDetChannels = GetAuxDetChannels(InputHitVector);
  std::vector<sim::AuxDetSimChannel> auxDetVector;
  auxDetVector.reserve(size(auxDetChannels));

  for (auto const channelNum : auxDetChannels) {
    auxDetVector.push_back(GetAuxDetSimChannelByNumber(InputHitVector, channelNum));
  }

  return auxDetVector;
}
```

Reply:
Changes made